noise_corrector.py

```diff
@@ -71,149 +207,229 @@ def estimate_true_values(observed_values, randomized_response_rate, beta=0):
-  # individual true report. However, we can optimize this by directly computing
-  # the number of total responses that would have flipped, and uniformly
-  # distributing them across all k buckets.
-  x = randomized_response_rate
-
-  # Compute the non-flipped counts.
-  non_flipped_counts = [numpy.random.binomial(true_count, 1 - x) for true_count in true_reports]
-
-  # Distribute the flipped reports uniformly among all k buckets.
-  num_flipped = n - sum(non_flipped_counts)
-  flipped_counts = numpy.random.multinomial(num_flipped, [1/k] * k)
-  return non_flipped_counts + flipped_counts
-
-if __name__ == "__main__":
-  run_example()
+    return (v - n * x / k) / (1 - x)
+
+  return {k: estimate(v) for k, v in observed_values.items()}
+
+
+def join_reports_with_sources(sources: List[Source], reports: List[Report]) -> JoinedList:
+  '''Returns tuple of (source, [reports]) joined with the source_event_id'''
+  report_map = collections.defaultdict(list)
+  for r in reports:
+    report_map[r['report']['source_event_id']].append(r)
+
+  def join(source) -> Joined:
+    source_event_id = source['registration_config']['source_event_id']
+    return (source, report_map[source_event_id])
+  return [join(s) for s in sources]
+
+
+################## AGGREGATE UTILITIES ##################
+
+def get_raw_corrected_map(joined: JoinedList, params: ParamConfig) -> Dict[OutputEnumeration, float]:
+  '''Returns an aggregate map of OutputEnumeration -> float, with noise debiased
+  according to params'''
+  output_counts = {o: 0 for o in OutputEnumeration.generate_all(params)}
+  for j in joined:
+    output = OutputEnumeration.create_from_data(j)
+    assert output in output_counts
+    output_counts[output] += 1
+
+  return estimate_true_values(output_counts, params.rr_rate)
+
+
+def correct_aggregates(joined: JoinedList, params: ParamConfig) -> List[dict]:
+  '''Returns debiased report counts broken out by trigger_data'''
+  corrected = get_raw_corrected_map(joined, params)
+  histogram = numpy.zeros(params.data_cardinality)
+  for o, c in corrected.items():
+    histogram += c * o.data_histogram()
+  x = 10
+  return [{'trigger_data': t,
+           'report_count': v} for t, v in enumerate(histogram)]
+
+
+################## EVENT-LEVEL UTILITIES ##################
+
+def adjust_to_match_distribution(values: List[T],
+                                 estimated_counts: Dict[T, float],
+                                 default_value: T) -> List[T]:
+  '''Given a list of values, and an aggregate map that counts all possible values,
+  returns a list of adjusted values to match the aggregate distribution.
+
+  values: list of arbitrary objects (must be possible to place in dict/set). To
+          minimize bias, this list should be shuffled prior to calling this
+          method.
+  estimated_counts: map of value -> estimated count. All values in `values` must be
+                    present. Note that negatives are allowed, but it will mean the
```

Comment:
I think this part is slightly misleading, at least as far as I understand the "bias claim". My understanding (and please correct me if I'm wrong) is that we view the entire algorithm, including randomization, debiasing the aggregates, and the adjustment, as a single process. In this case, the end value will always be biased since we are not allowed to assign anything less than zero. As a result, it seems a bit misleading to say that it is only biased for the negative estimates case. Wdyt?

Reply:
I think this is true, I've tried rewording it to match your comment:

```
Note that negatives are allowed, but they will be
clamped to 0 internally. Note that because the we cannot
assign any output less than 0, the end values will always be
biased
```
Does that make sense to you?

Reply:
Yes the new explanation LGTM. Thanks!